### PR TITLE
FP16 resurrected

### PR DESCRIPTION
**Pull request by benson31**
A while back I had started some half-precision stuff. Much was lost in a restore-from-backup fail (slash I-stupidly-never-pushed-that-branch-out-to-GH-and-then-my-kernel-died fail), but I was able to find a few pieces lying around. Much has changed in Hydrogen since the first work, so this resurrects the pieces I've found and updates things so that the `Matrix<half_float::half, Device::CPU>` will compile. @gunney1 this is a good place to start.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -134,6 +134,20 @@ endif ()
 # MEMORY-RELATED OPTIONS
 #
 
+option(${PROJECT_NAME}_ENABLE_HALF
+  "Enable the use of \"half\" library." OFF)
+# This refers to the half.sourceforge.net type "half"
+
+option(${PROJECT_NAME}_ENABLE_BFLOAT32
```
Comment (ndryden) [inline CMakeLists.txt]: Do you mean bfloat16?

**Review by ndryden — APPROVED** [review]
This seems like a good start on FP16 as a storage format. Doesn't seem to have extensive support for computation with it (unless I've missed something).

File: CMakeLists.txt (hunk 1 above)
Comment (benson31) [reply CMakeLists.txt]: _sigh_. Yes. Yes I do.

**Comment by benson31** [comment]
There is no computation done here. Because of the "accumulate at fp32" bit, we need to make sure to be capturing all appropriate accumulation before returning to this storage format. Since this really only does stuff on CPU, anyway, I was mostly envisioning the CPU matrix as storage-only and compute would be offloaded to GPU, at least until the CPU BLAS libraries catch up/HW support for at least some FP16 types becomes available.

File: cmake/configure_files/hydrogen_config.h.in [hunk 2]
```diff
@@ -11,7 +11,9 @@
 #cmakedefine HYDROGEN_BLAS_SUFFIX @HYDROGEN_BLAS_SUFFIX@
 #cmakedefine HYDROGEN_LAPACK_SUFFIX @HYDROGEN_LAPACK_SUFFIX@
 
-// MKL stuff
+// 16-bit floating point stuff
+#cmakedefine HYDROGEN_HAVE_HALF
+#cmakedefine HYDROGEN_HAVE_BFLOAT32
```
Comment (ndryden) [inline cmake/configure_files/hydrogen_config.h.in]: More `BFLOAT32`.

**Review by naoyam — APPROVED** [review]
Looks good to me.

**Review by timmoon10 — APPROVED** [review]
LGTM

**Pull request merged**